File: experiments/runs/2026-09-14_17-58-27_claim-office-example-mapping_external-tcr-kentbeck-2026-09-14-cc_opus-5-requesty-no-thinking/src/policy.ts
import { Item } from './premium.js';
import { insuranceSum } from './quote.js';
import { damagePayout } from './claim.js';

const CAP_FACTOR = 2;

export interface Damage {
  itemType: string;
  amount: number;
}

export interface ClaimResult {
  payout: number;
  remainingCap: number;
}

export class Policy {
  private readonly items: Item[];
  private remaining: number;

  constructor(items: Item[]) {
    this.items = items;
    this.remaining = insuranceSum(items) * CAP_FACTOR;
  }

  get remainingCap(): number {
    return this.remaining;
  }

  claim(damages: Damage[]): ClaimResult {
    const available = new Map<string, Item[]>();
    for (const item of this.items) {
      const bucket = available.get(item.type) ?? [];
      bucket.push(item);
      available.set(item.type, bucket);
    }

    let gross = 0;
    for (const damage of damages) {
      if (damage.amount < 0) {
        throw new Error(`negative damage amount: ${damage.amount}`);
      }
      const bucket = available.get(damage.itemType);
      if (!bucket || bucket.length === 0) {
        throw new Error(`item not covered by the policy: ${damage.itemType}`);
      }
      const item = bucket.shift()!;
      gross += damagePayout(item, damage.amount);
    }

    const payout = Math.min(Math.floor(gross), this.remaining);
    this.remaining -= payout;
    return { payout, remainingCap: this.remaining };
  }
}

File: experiments/runs/2026-09-20_06-48-34_claim-office-java-example-mapping_baseline-inline-tdd-v1-cc_opus-5-no-thinking/src/main/java/Policy.java
import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.ArrayList;
import java.util.List;

/** A policy created by a quote step; it tracks the cap consumed by earlier claims. */
public class Policy {

    private static final BigDecimal DEDUCTIBLE = BigDecimal.valueOf(100);
    private static final BigDecimal HALF = new BigDecimal("0.5");
    private static final int HALVING_ENCHANTMENT_LEVEL = 8;
    private static final int CAP_FACTOR = 2;

    private final List<Item> items;
    private BigDecimal remainingCap;

    public Policy(List<Item> items) {
        this.items = List.copyOf(items);
        this.remainingCap = BigDecimal.valueOf(CAP_FACTOR * insuranceSum());
    }

    /** The sum of the items' unmodified insurance values; discounts and surcharges do not affect it. */
    private long insuranceSum() {
        return items.stream().mapToLong(item -> PriceList.insuranceValue(item.type())).sum();
    }

    public int remainingCap() {
        return remainingCap.setScale(0, RoundingMode.FLOOR).intValueExact();
    }

    public ClaimResult claim(List<Damage> damages) {
        List<Item> available = new ArrayList<>(items);
        BigDecimal desired = BigDecimal.ZERO;
        for (Damage damage : damages) {
            desired = desired.add(reimbursement(damage, take(available, damage.itemType())));
        }
        BigDecimal payout = desired.min(remainingCap);
        remainingCap = remainingCap.subtract(payout);
        return new ClaimResult(payout.setScale(0, RoundingMode.FLOOR).intValueExact(), remainingCap());
    }

    /**
     * The reimbursement for a single damage entry: the special clauses reduce the damage
     * amount first, then the entry's own deductible applies.
     */
    private BigDecimal reimbursement(Damage damage, Item item) {
        if (damage.amount() < 0) {
            throw new ClaimOfficeException("damage amount must not be negative: " + damage.amount());
        }
        BigDecimal amount = BigDecimal.valueOf(damage.amount());
        if (item.enchantment() >= HALVING_ENCHANTMENT_LEVEL) {
            amount = amount.multiply(HALF);
        }
        return amount.subtract(DEDUCTIBLE).max(BigDecimal.ZERO);
    }

    /**
     * Consumes one insured item of the given type, so that a policy covering two swords can
     * absorb two sword damages while a single sword cannot.
     */
    private Item take(List<Item> available, String itemType) {
        for (int i = 0; i < available.size(); i++) {
            if (available.get(i).type().equals(itemType)) {
                return available.remove(i);
            }
        }
        throw new ClaimOfficeException("item not covered by the policy: " + itemType);
    }
}
